feat(google-analytics): Enhance GA4 integration

BrocksiNet opened the pull request:
### 1. Why is this change necessary?

The existing Google Analytics integration was not fully compliant with GA4 e-commerce standards. Key data points such as `currency`, `value`, `item_brand`, and hierarchical category structures were missing from tracking events. This prevented complete and accurate e-commerce funnel analysis in Google Analytics 4. Additionally, important user interaction events were not being tracked.

### 2. What does this change do, exactly?

This change refactors the Storefront's Google Analytics integration to be fully GA4 compliant.

**New Events Added:**
- `add_to_wishlist` - Tracks when users add products to their wishlist
- `remove_from_wishlist` - Tracks when users remove products from their wishlist
- `view_cart` - Tracks when users view their shopping cart
- `add_shipping_info` - Tracks selected shipping method during checkout
- `add_payment_info` - Tracks selected payment method during checkout

**Enriched Existing Events with GA4 Parameters:**
- `view_item` - Added `currency`, `value`, `item_brand`, hierarchical `item_category`
- `view_item_list` - Added `currency`, `value`, `item_category` for all items
- `add_to_cart` - Added `currency`, `value`, `item_brand`, hierarchical `item_category`
- `begin_checkout` - Added `currency`, `value`
- `remove_from_cart` - Added `currency`, `value`
- `purchase` - Already compliant, verified

**Pagination Support:**
- `view_item_list` and `add_to_cart` events now fire correctly after pagination/filtering on listing pages
- Events subscribe to `Listing/afterRenderResponse` to track interactions on dynamically loaded content

**New Helper Classes:**
- `ProductPageHelper` - Extracts product data from DOM (meta tags, breadcrumbs, product cards)
- `LineItemHelper.getCategoriesFromElement()` - Shared category extraction logic
- `EventAwareAnalyticsEvent._getPluginInstances()` - Shared plugin instance retrieval

**New Configuration:**
- `core.cart.openOffcanvasAfterAddToCart` - Controls whether offcanvas cart opens after adding products (default: `true`)
- Renamed from `disableOffcanvasAfterAddToCart` for better UX (positive phrasing instead of double-negative)

**Breaking Change - Deprecated Event:**
- ⚠️ The `checkout_progress` event is **no longer fired**. This was a Universal Analytics (GA3) event that has been replaced by the GA4-specific events: `view_cart`, `add_shipping_info`, and `add_payment_info`. The event class remains for backwards compatibility but will be removed in v6.8.0.

**Data Persistence:**
- Brand and category information is now captured when adding products to cart via the line item `payload`
- Data is available for all subsequent cart and checkout events

**Code Quality:**
- Added unit tests for all GA4 events
- Refactored to use shared helpers (DRY principle)
- Updated ESLint compliance (dot notation, template literals, Object.hasOwn)

### 3. Describe each step to reproduce the issue or behaviour.

1. Configure a Google Analytics tracking ID in the Administration
2. Use a browser extension (e.g., Google Analytics Debugger) to monitor `gtag` events
3. Navigate the storefront and perform these actions:
   - View a category/listing page → verify `view_item_list` has `currency`, `value`, `item_category`
   - **Use pagination or filters** → verify `view_item_list` fires again with updated items
   - View a product detail page → verify `view_item` has all GA4 parameters
   - Add a product to the wishlist → verify `add_to_wishlist` fires
   - Remove from wishlist → verify `remove_from_wishlist` fires
   - Add a product to the cart → verify `add_to_cart` has all parameters including `value` and `brand`
   - **Add to cart from paginated listing** → verify `add_to_cart` fires with correct data
   - Open the cart page → verify `view_cart` fires with `currency` and `value`
   - Click checkout → verify `begin_checkout` has `currency` and `value`
   - On confirm page → verify `add_shipping_info` and `add_payment_info` fire
   - Complete purchase → verify `purchase` event fires

### 4. Please link to the relevant issues (if any).

- closes #9813
- closes #10863
- closes #14136
- closes #14137
- closes #14138
- closes #14139

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have updated `RELEASE_INFO-6.7.md` with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

BrocksiNet committed:
feat(google-analytics): Enhance GA4 integration with new events and enriched data parameters

- Updated Google Analytics integration to comply with GA4 standards, improving e-commerce tracking.
- Added new events: `add_to_wishlist` and `view_cart`.
- Enriched existing events (`add_to_cart`, `begin_checkout`, `purchase`, `view_item`, `remove_from_cart`) with additional parameters including `currency`, `value`, `item_brand`, and hierarchical `item_category`.
- Adjusted payload handling in various controllers and helper functions to support new data structure.

BrocksiNet committed:
feat(google-analytics): fix twigLint and jest failing test

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat(google-analytics): Avoid console errors if plugin instance is not found

BrocksiNet committed:
fix(google-analytics): Update category key generation for GA4 compliance

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
fix(google-analytics): update begin-checkout event handling to ensure button event listeners are managed correctly

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat(google-analytics): enhance GA4 integration with new events and data tracking

- Updated Google Analytics integration to comply with GA4 standards, adding new events: `add_payment_info`, `add_shipping_info`, `remove_from_wishlist`, and improved tracking for existing events.
- Removed deprecated `checkout_progress` event, ensuring proper tracking flow with `view_cart` and other GA4 events.
- Introduced `ProductPageHelper` for better product data extraction across various pages.
- Enhanced event data with additional parameters such as `currency`, `value`, `item_brand`, and hierarchical `item_category` structure for improved e-commerce tracking.
- Updated tests to cover new event functionalities and ensure accurate data handling.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
fix: handle undefined manufacturer and category names in buy widget and checkout components

- Updated buy-widget-form.html.twig to safely access manufacturer names, ensuring no errors occur when the manufacturer is undefined.
- Enhanced hidden-line-items-information.html.twig to check for the existence of category names before slicing, preventing potential issues with undefined values.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat: add configuration options for offcanvas cart tracking

- Introduced `Track offcanvas cart` setting in Sales Channel Analytics to fire `view_cart` events when the offcanvas cart is opened.
- Added `Disable automatic offcanvas cart after adding a product` option to prevent the offcanvas from opening automatically, showing a success message instead.
- Updated Google Analytics integration to support enhanced tracking for offcanvas cart interactions.
- Refactored related components and tests to accommodate new features and ensure accurate funnel tracking.

BrocksiNet committed:
test: add unit tests for Migration1765205483AddTrackOffcanvasCartToAnalytics

- Implemented tests to verify the creation timestamp and migration process for the new `track_offcanvas_cart` column in the `sales_channel_analytics` table.
- Included rollback functionality to ensure the test environment is reset before each test run.

BrocksiNet commented:
### Two new configuration options

#### Enable/disable offcanvas cart tracking:
<img width="1592" height="932" alt="shopware-analytics-track-offcanvas-cart-on-off" src="https://github.com/user-attachments/assets/761aec33-dc9e-4956-a5be-e5fb4c2deb0e" />

#### Disable the offcanvas cart opening after a product is added to the cart:
<img width="1546" height="946" alt="shopware-cart-new-setting-offcanvas-on-off" src="https://github.com/user-attachments/assets/c6b89728-f8e2-48e3-bf45-78fcfa8451a0" />

BrocksiNet committed:
test: add unit test for addLineItems with nested array payload

- Implemented a new test to verify that the addLineItems method correctly handles a line item with a nested array payload.
- Ensured that the expected line item data is created and passed to the line item registry during the addLineItems process.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat(analytics): Add track_offcanvas_cart field to sales channel analytics

BrocksiNet committed:
refactor: Enhance payload handling in cart and analytics plugins

- Updated ProductLineItemFactory to ensure payload is an array before merging.
- Refactored CartLineItemController to normalize payloads using a dedicated method.
- Improved error handling in AddToCartPlugin for fetch failures, falling back to offcanvas.
- Added LineItemHelper to retrieve product data from hidden line items, enhancing analytics events for add/remove from wishlist.
- Introduced tests for new functionalities in line item helper and wishlist events.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

patzick reviewed (commented):
My remarks regarding the modified schema:

1. in responses this field should be marked as required
2. while creation/updates it can be marked as not required
3. Additionally we have circular dependency as `SalesChannelAnalytics` <-> `SalesChannel`, something's off here

Marking it always "optional" makes client not know really when to expect and rely on that field

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat: add default value for trackOffcanvasCart in SalesChannelAnalyticsDefinition

BrocksiNet commented:
The field now has the `Required` flag with a default value in `getDefaults()`, ensuring it's NOT NULL in the database with a default of `false`, making it optional during creation/updates while always present in responses. Unfortunately, Shopware's OpenAPI schema generation creates a single schema for both reads and writes. Fields with defaults are excluded from the `required` array to allow optional writes, so we cannot mark them as "required in responses only" without a custom schema override. The circular dependency between `SalesChannelAnalytics` and SalesChannel is a standard bidirectional one-to-one relationship pattern and not an issue.

If we want to have a different response Schema for Entities, we somehow need to solve this in a dedicated task.

patzick commented:
Then it's fine to leave it like this and follow up in https://github.com/shopware/shopware/issues/14039 as this will also touch this schema as well :)

patzick reviewed (approved):
From schema perspective 👍

BrocksiNet committed:
feat: add Sales Channel Analytics API schema and endpoints

BrocksiNet committed:
refactor: update SalesChannelAnalytics schema to use union types for nullable fields

BrocksiNet committed:
refactor: remove path, only schema with required fields

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
feat: enhance Google Analytics events for cart and item list interactions

- Added debouncing to the ViewCartEvent to prevent duplicate events during rapid cart updates.
- Updated ViewCartEvent to subscribe to both 'offCanvasOpened' and 'registerEvents' for better tracking of cart changes.
- Refactored ViewItemListEvent to extend EventAwareAnalyticsEvent and added subscription to 'Listing/afterRenderResponse' for AJAX updates.
- Improved tests for both events to cover new functionality and ensure correct event firing on updates.

BrocksiNet committed:
docs: update RELEASE_INFO-6.7 for Google Analytics enhancements

- Added details on the `view_item_list` event firing during pagination and filter application.
- Expanded description of the `Track offcanvas cart` configuration to include updates on cart content.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -59,12 +59,7 @@ When no `Accept` header is set or with `*/*`, `PDF` will be returned. (PR #12944
 
-### PHP 8.5 support
-
-Shopware is now fully compatible with PHP 8.5.
```
merge foo :see_no_evil:

BrocksiNet replied:
Gone after merging trunk ...

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

tobiasberge commented on src/Core/System/Resources/config/cart.xml:
```diff
@@ -17,6 +17,14 @@
+            <label>Disable automatic offcanvas cart after adding a product</label>
+            <label lang="de-DE">Automatischen Offcanvas-Warenkorb nach dem Hinzufügen eines Produkts deaktivieren</label>
+            <helpText>When enabled, the offcanvas cart will NOT automatically open after adding a product. Instead, only a success message will be displayed. Enable this if customers tend to add many products to their shopping cart.</helpText>
+            <helpText lang="de-DE">Wenn aktiviert, wird der Offcanvas-Warenkorb NICHT automatisch geöffnet, nachdem ein Produkt hinzugefügt wurde. Stattdessen wird nur eine Erfolgsmeldung angezeigt. Aktivieren Sie dies wenn Kunden tendenziell viele Produkte dem Warenkorb hinzufügen.</helpText>
```
Would it also be possible to turn this setting into `openOffcanvasAfterAddToCart` which is `true` by default? I'm not the biggest fan of these "negated" settings tbh, like "turn this ON in order to turn something OFF" instead of "you can now turn this OFF by turning it OFF. But it is ON by default". We also write "NOT" in all-caps in the helptext to avoid confusion already.

tobiasberge replied:
Probably this is technically safer because we don't rely immediately on a newly introduced config field. Was just speaking from a UX perspective.

BrocksiNet replied:
I switched that config and added a migration to set the default to true. ✅

tobiasberge commented on src/Core/System/Resources/config/cart.xml:
```diff
@@ -17,6 +17,14 @@
         </input-field>
 
+        <input-field type="bool">
+            <name>disableOffcanvasAfterAddToCart</name>
```
Nice 👍 That we have a setting for this now.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/google-analytics/events/remove-from-cart.event.js:
```diff
@@ -23,14 +24,55 @@ export default class RemoveFromCart extends AnalyticsEvent
+        }
+
+        const categories = {};
+        for (let i = 1; i <= 5; i++) {
```
Is there a reason thy this is hard coded to 5 categories?

tobiasberge replied:
Ahh nevermind, here the explanation. 

> {# Google Analytics supports up to 5 category levels. We take the first 5 from the breadcrumb. #}

Maybe we can add this comment here as well.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/google-analytics/events/remove-from-wishlist.event.js:
```diff
@@ -0,0 +1,109 @@
+        // Fallback to line item data (cart/checkout/finish pages)
+        if (!productData.name) {
+            const lineItemData = LineItemHelper.getProductData(productId);
+            if (lineItemData) {
```
Maybe like this? But not sure if the query for `LineItemHelper.getProductData(productId)` works without the `!productData.name` check.  

```js
const lineItemData = LineItemHelper.getProductData(productId);

if (!productData.name && lineItemData) {
  // things
}

```

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/add-to-cart/add-to-cart.plugin.js:
```diff
@@ -80,7 +86,88 @@ export default class AddToCartPlugin extends Plugin {
+    _shouldOpenOffcanvas() {
+        const flag = window.disableOffcanvasAfterAddToCart;
+
+        if (flag === undefined) {
```
Can this ever be undefined? Because the window prop seems to be always in the template now. And `view-cart.event.js‎` -> `_isOffcanvasTrackingEnabled` also uses a one-liner to check a similar window prop. But I'm probably missing something here. Was just wondering.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/google-analytics/events/remove-from-wishlist.event.js:
```diff
@@ -0,0 +1,109 @@
+            return;
+        }
+
+        const instances = typeof plugin.get === 'function' ? plugin.get('instances') : null;
```
Are we not able to use `window.PluginManager.getPluginInstances()` here like in other places? This checks seem a bit cumbersome.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/google-analytics/product-page.helper.js:
```diff
@@ -0,0 +1,113 @@
+     * @returns {string|undefined}
+     */
+    static getBrand() {
+        return document.querySelector('div[itemprop="brand"] meta[itemprop="name"]')?.content;
```
Would this selectors also work without the element selectors?

```
'[itemprop="brand"] [itemprop="name"]'
```

In case we have to change an element type in the future we don't have a BC.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/google-analytics/product-page.helper.js:
```diff
@@ -0,0 +1,113 @@
+     * @returns {Object}
+     */
+    static getCategories() {
+        const breadcrumbNodes = document.querySelectorAll('nav[aria-label="breadcrumb"] .breadcrumb-title');
```
Same here

tobiasberge commented on src/Storefront/Resources/views/storefront/component/checkout/add-to-cart-alert-template.html.twig:
```diff
@@ -0,0 +1,16 @@
+#}
+
+{% block component_add_to_cart_alert_template %}
+    <div class="alert alert-success fade d-flex align-items-center d-none js-add-to-cart-alert-template mt-2 mb-0 py-2"
```
The JS part looks like its cloning this node and injects it into the DOM. Which is good because it should trigger the screen reader because of the `role="alert"`. When this is supposed to be a template for JS only maybe we can wrap it into an actual `<template>` tag.

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/add-to-cart/add-to-cart.plugin.js:
```diff
@@ -80,7 +86,88 @@ export default class AddToCartPlugin extends Plugin {
+
+        setTimeout(() => {
+            alert.addEventListener('transitionend', () => alert.remove(), { once: true });
+            alert.classList.remove('show');
```
Could we do this via the Bootstrap alert methods? https://getbootstrap.com/docs/5.3/components/alerts/#methods

Not sure, its a bit weird since they seem to only cover the "close" logic but do not have a "show" case.

BrocksiNet replied:
I would like to stick with the vanilla approach in that case. 
The alert does not fit because it is missing the "show" part.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

BrocksiNet committed:
refactor: update AddToCart plugin and Google Analytics event handling

- Changed the way the AddToCart alert is created to use a template element for better structure.
- Enhanced EventAwareAnalyticsEvent to manage dynamic content and prevent duplicate subscriptions.
- Updated AddToCartEvent to utilize product data from the detail page or fallback to product card data.
- Adjusted ProductPageHelper to improve selector queries for brand and category retrieval.
- Added tests for AddToCartEvent to ensure proper event firing and subscription behavior.

BrocksiNet committed:
refactor(google-analytics): enhance plugin instance handling for optional plugins

- Updated EventAwareAnalyticsEvent to clarify that the absence of a plugin is optional.
- Refactored RemoveFromWishlistEvent to streamline the retrieval of WishlistStorage plugin instances, ensuring it handles optional plugins gracefully.

BrocksiNet committed:
refactor(google-analytics): streamline product data retrieval in RemoveFromWishlistEvent

- Simplified the logic for retrieving product data by consolidating the fallback mechanism for line item data.
- Ensured that line item data is only accessed when product data is not available, improving code clarity.

BrocksiNet committed:
refactor(google-analytics): extract category data retrieval into LineItemHelper

- Introduced a new static method in LineItemHelper to extract GA4 category data from element attributes, supporting up to 5 category levels.
- Updated RemoveFromCart event to utilize the new method for improved code clarity and reusability.

BrocksiNet committed:
refactor(add-to-cart): simplify offcanvas opening logic

- Streamlined the logic in the _shouldOpenOffcanvas method to directly return the condition for opening the offcanvas, improving code readability.

BrocksiNet committed:
feat(cart): add configuration for opening offcanvas cart after adding products

- Introduced a new migration to add the configuration setting 'core.cart.openOffcanvasAfterAddToCart' to control the behavior of the offcanvas cart.
- Updated the cart XML configuration to reflect the new setting and its default value.
- Modified the AddToCartPlugin to utilize the new configuration for determining whether to open the offcanvas cart.
- Adjusted related tests to ensure proper functionality with the new setting.
- Updated Twig templates to check the new configuration instead of the old one.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

tobiasberge commented on RELEASE_INFO-6.7.md:
```diff
@@ -197,6 +192,33 @@ In the settings module, there is now a search bar in the top right. It can be us
+
+#### New Configuration: Disable Automatic Offcanvas Cart
+
+A new configuration option `Disable automatic offcanvas cart after adding a product` has been added to the Cart settings (Settings → Shop → Cart). When enabled:
```
Looks like this still needs to be adjusted due to the changed setting.

BrocksiNet committed:
chore(release): update RELEASE_INFO-6.7 with GA4 event enhancements and new cart configurations

- Updated Google Analytics integration details to reflect changes in event tracking, including the firing of `view_item_list` and `add_to_cart` events during pagination and filtering.
- Added new configuration options for tracking offcanvas cart interactions and controlling the automatic opening of the offcanvas cart after adding products, with updated recommendations for accurate funnel tracking.

BrocksiNet committed:
Merge branch 'trunk' into feat/google-analytics-4-update

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -133,12 +133,7 @@ When no `Accept` header is set or with `*/*`, `PDF` will be returned. (PR #12944
 
 ## Core
 
-### PHP 8.5 support
```
why did you remove this?

BrocksiNet replied:
Good question, added back ... and fixed positions for new release version

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -197,6 +192,33 @@ In the settings module, there is now a search bar in the top right. It can be us
 
 ## Storefront
 
+### Google Analytics 4 Integration Update
```
this needs to be moved to the next minor version section. it is in the 6.7.6.0 section, where it will not be part of

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -197,6 +192,33 @@ In the settings module, there is now a search bar in the top right. It can be us
+
+**Recommended for accurate funnel tracking:** Disable "Open offcanvas cart after adding a product" and enable "Track offcanvas cart". This ensures `view_cart` events only fire when users intentionally click the cart button, providing accurate funnel metrics.
+
+#### Deprecated `checkout_progress` Event
```
if you deprecate something, you should already document the removal in the 6.8 upgrade guide

BrocksiNet committed:
chore(release): update RELEASE_INFO-6.7 and UPGRADE-6.8 with GA4 enhancements and deprecations
